File: tool/cleanup_cart.js
const admin = require('firebase-admin');
const { getFirestore } = require('firebase-admin/firestore');

// Initialize Firebase Admin
admin.initializeApp({
  projectId: 'tripbasket-sctkxj'
});

const db = getFirestore();

async function cleanupCart() {
  console.log('🧹 Starting cart cleanup...');
  
  try {
    // Get all cart items
    const cartSnapshot = await db.collection('cart').get();
    console.log(`📦 Found ${cartSnapshot.docs.length} cart items`);
    
    let deletedCount = 0;
    let validCount = 0;
    
    for (const cartDoc of cartSnapshot.docs) {
      const cartData = cartDoc.data();
      const tripReference = cartData.tripReference;
      
      if (!tripReference) {
        console.log(`❌ Cart item ${cartDoc.id} has null trip reference, deleting...`);
        await cartDoc.ref.delete();
        deletedCount++;
        continue;
      }
      
      // Check if the referenced trip still exists
      try {
        const tripDoc = await tripReference.get();
        if (!tripDoc.exists) {
          console.log(`❌ Trip ${tripReference.id} no longer exists, deleting cart item ${cartDoc.id}...`);
          await cartDoc.ref.delete();
          deletedCount++;
        } else {
          console.log(`✅ Cart item ${cartDoc.id} references valid trip ${tripReference.id}`);
          validCount++;
        }
      } catch (e) {
        console.log(`❌ Error checking trip ${tripReference.id}, deleting cart item ${cartDoc.id}: ${e.message}`);
        await cartDoc.ref.delete();
        deletedCount++;
      }
      
      // Small delay to avoid hitting rate limits
      await new Promise(resolve => setTimeout(resolve, 100));
    }
    
    console.log('🎉 Cleanup completed!');
    console.log(`✅ Valid cart items: ${validCount}`);
    console.log(`❌ Deleted orphaned items: ${deletedCount}`);
    
  } catch (e) {
    console.log(`💥 Error during cleanup: ${e.message}`);
  }
  
  process.exit(0);
}

cleanupCart();

File: tool/cleanup_cart.dart
import 'dart:async';
import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:firebase_core/firebase_core.dart';
import '../lib/firebase_options.dart';

Future<void> main() async {
  // Initialize Firebase
  await Firebase.initializeApp(
    options: DefaultFirebaseOptions.currentPlatform,
  );

  final firestore = FirebaseFirestore.instance;
  
  print('🧹 Starting cart cleanup...');
  
  try {
    // Get all cart items
    final cartSnapshot = await firestore.collection('cart').get();
    print('📦 Found ${cartSnapshot.docs.length} cart items');
    
    int deletedCount = 0;
    int validCount = 0;
    
    for (final cartDoc in cartSnapshot.docs) {
      final cartData = cartDoc.data();
      final tripReference = cartData['tripReference'] as DocumentReference?;
      
      if (tripReference == null) {
        print('❌ Cart item ${cartDoc.id} has null trip reference, deleting...');
        await cartDoc.reference.delete();
        deletedCount++;
        continue;
      }
      
      // Check if the referenced trip still exists
      try {
        final tripDoc = await tripReference.get();
        if (!tripDoc.exists) {
          print('❌ Trip ${tripReference.id} no longer exists, deleting cart item ${cartDoc.id}...');
          await cartDoc.reference.delete();
          deletedCount++;
        } else {
          print('✅ Cart item ${cartDoc.id} references valid trip ${tripReference.id}');
          validCount++;
        }
      } catch (e) {
        print('❌ Error checking trip ${tripReference.id}, deleting cart item ${cartDoc.id}: $e');
        await cartDoc.reference.delete();
        deletedCount++;
      }
    }
    
    print('🎉 Cleanup completed!');
    print('✅ Valid cart items: $validCount');
    print('❌ Deleted orphaned items: $deletedCount');
    
  } catch (e) {
    print('💥 Error during cleanup: $e');
  }
  
  exit(0);
}